llap-server/src/test/org/apache/hadoop/hive/llap/daemon/impl/TestTaskExecutorService.java

```diff
@@ -164,73 +165,84 @@ private void testPreemptionHelper(
+
 
     TaskExecutorServiceForTest taskExecutorService =
-        new TaskExecutorServiceForTest(1, 2, ShortestJobFirstComparator.class.getName(), true, mockMetrics);
+        new TaskExecutorServiceForTest(2, 2, ShortestJobFirstComparator.class.getName(), true, mockMetrics);
     taskExecutorService.init(new Configuration());
     taskExecutorService.start();
 
     try {
-      String fragmentId = r1.getRequestId();
-      Scheduler.SubmissionState submissionState = taskExecutorService.schedule(r1);
-      assertEquals(Scheduler.SubmissionState.ACCEPTED, submissionState);
+      String fragmentId1 = r1.getRequestId();
+      Scheduler.SubmissionState submissionState1 = taskExecutorService.schedule(r1);
+      assertEquals(Scheduler.SubmissionState.ACCEPTED, submissionState1);
       awaitStartAndSchedulerRun(r1, taskExecutorService);
 
+      String fragmentId2 = r2.getRequestId();
+      Scheduler.SubmissionState submissionState2 = taskExecutorService.schedule(r2);
+      assertEquals(Scheduler.SubmissionState.ACCEPTED, submissionState2);
+      awaitStartAndSchedulerRun(r2, taskExecutorService);
+
       TaskWrapper taskWrapper = taskExecutorService.preemptionQueue.peek();
       assertNotNull(taskWrapper);
       assertTrue(taskWrapper.isInPreemptionQueue());
+      assertEquals(fragmentId1, taskWrapper.getRequestId());
+      assertEquals(2, taskExecutorService.preemptionQueue.size());
 
       // Now notify the executorService that the task has moved to finishable state.
       r1.setCanUpdateFinishable();
-      taskWrapper.finishableStateUpdated(true);
-      TaskWrapper taskWrapper2 = taskExecutorService.preemptionQueue.peek();
-      assertNotNull(taskWrapper2);
-      assertTrue(taskWrapper.isInPreemptionQueue());
+      r2.setCanUpdateFinishable();
 
-      // And got a duck.
-      boolean result = taskExecutorService.updateFragment(fragmentId, true);
-      assertTrue(result);
-      taskWrapper2 = taskExecutorService.preemptionQueue.peek();
-      assertNull(taskWrapper2);
-      assertFalse(taskWrapper.isInPreemptionQueue());
+      TaskWrapper taskWrapper1 = taskExecutorService.knownTasks.get(fragmentId1);
+      TaskWrapper taskWrapper2 = taskExecutorService.knownTasks.get(fragmentId2);
 
-      r1.complete();
-      r1.awaitEnd();
-
-      // Now start with everything and test losing stuff.
-      fragmentId = r2.getRequestId();
-      submissionState = taskExecutorService.schedule(r2);
-      assertEquals(Scheduler.SubmissionState.ACCEPTED, submissionState);
-      awaitStartAndSchedulerRun(r2, taskExecutorService);
+      // r1 is guaranteed now, so only r2 exists in pre-emption queue
+      taskExecutorService.updateFragment(fragmentId1, true);
+      taskWrapper = taskExecutorService.preemptionQueue.peek();
+      assertNotNull(taskWrapper);
+      assertTrue(taskWrapper.isInPreemptionQueue());
+      assertEquals(fragmentId2, taskWrapper.getRequestId());
+      assertEquals(1, taskExecutorService.preemptionQueue.size());
```

Comment:
Do we assume that r1 would be in the wait queue now?
Shall we check this?
Not sure who would be responsible to move the task from the preemptionQ to the waitQ